File: CSharpCodeBase/entities/builderentity/builderentity.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class builderentity :entity{
 public void init (self);
       Entity.init(self);
       this.enabled = true;
       this.active = false;
       this.static = false;
       this.doneCallback = null;
       this.interruptCallback = null;
 }})
 public void StartConstruction(entityName, ){neCallback, interruptCallback){
   this.entity = GameController.entityFactory:Create(entityName, Input.RaycastMouseOnTerrain());
   this.entity:Enable();
   this.active = true;
   this.doneCallback = ){neCallback
   this.interruptCallback = interruptCallback;
 }
 public void InterruptConstruction(){
   this.entity:Disable();
   this.active = false;
   if(this.interruptCallback  ){ 
       self:interruptCallback();
     }
 }
 public void DoneConstruction(){
     GameController.entityFactory:CreateInWorld("TreeFirEntity", this.entity:GetPosition());
     this.entity:Disable();
     this.active = false;
     if(this.doneCallback  ){ 
       self:doneCallback();
     }
 }
 public void Update(){
   Entity.Update(self);
   if(this.active  ){ 
     if(Input.GetKeyDown(KeyCode.Escape)  ){ 
       self:InterruptConstruction();
     }
     
     if(GameController.inputService:BottomButtonWasPressed()  ||  Input.GetKeyDown(KeyCode.KeypadEnter)  ){ 
       if(this.entity.count == 0  ){ 
         self:DoneConstruction();
       }
     }
     
     if(Input.GetKey(KeyCode.KeypadPlus)  ){ 
       //повернуть +
       this.entity:Rotate(1);
     }else{if Input.GetKey(KeyCode.KeypadMinus)  ){ 
       this.entity:Rotate(-1);
        //повернуть -
     }
     
     if(Input.GetKey(KeyCode.UpArrow)  ){ 
       this.entity:MoveUpTransform(5);
     }else{if Input.GetKey(KeyCode.DownArrow)  ){ 
       this.entity:MoveUpTransform(-5);
     }
     if(Input.GetKey(KeyCode.RightArrow)  ){ 
       this.entity:MoveRightTransform(5);
     }else{if Input.GetKey(KeyCode.LeftArrow)  ){ 
       this.entity:MoveRightTransform(-5);
     }
     
   }else{
     if(Input.GetKeyDown(KeyCode.B)  ){ 
       self:StartConstruction("MockupEntity");
     }
   }
 }
 }}

File: CSharpCodeBase/entities/builderentity/builderentity.lua
local Class = require 'utils.hump.class'
local Entity = require 'unity.entity'

local BuilderEntity = Class({__includes=Entity, init = function (self)
      Entity.init(self)
      self.enabled = true
      self.active = false
      self.static = false
      self.doneCallback = nil
      self.interruptCallback = nil
end})

function BuilderEntity:StartConstruction(entityName, doneCallback, interruptCallback)
  self.entity = GameController.entityFactory:Create(entityName, Input.RaycastMouseOnTerrain())
  self.entity:Enable()
  self.active = true
  self.doneCallback = doneCallback
  self.interruptCallback = interruptCallback
end

function BuilderEntity:InterruptConstruction()
  self.entity:Disable()
  self.active = false
  if self.interruptCallback then
      self:interruptCallback()
    end
end

function BuilderEntity:DoneConstruction()
    GameController.entityFactory:CreateInWorld('TreeFirEntity', self.entity:GetPosition())
    self.entity:Disable()
    self.active = false
    if self.doneCallback then
      self:doneCallback()
    end
end

function BuilderEntity:Update()
  Entity.Update(self)
  if self.active then
    if Input.GetKeyDown(KeyCode.Escape) then
      self:InterruptConstruction()
    end
    
    if GameController.inputService:BottomButtonWasPressed() or Input.GetKeyDown(KeyCode.KeypadEnter) then
      if self.entity.count == 0 then
        self:DoneConstruction()
      end
    end
    
    if Input.GetKey(KeyCode.KeypadPlus) then
      --повернуть +
      self.entity:Rotate(1)
    elseif Input.GetKey(KeyCode.KeypadMinus) then
      self.entity:Rotate(-1)
       --повернуть -
    end
    
    if Input.GetKey(KeyCode.UpArrow) then
      self.entity:MoveUpTransform(5)
    elseif Input.GetKey(KeyCode.DownArrow) then
      self.entity:MoveUpTransform(-5)
    end
    if Input.GetKey(KeyCode.RightArrow) then
      self.entity:MoveRightTransform(5)
    elseif Input.GetKey(KeyCode.LeftArrow) then
      self.entity:MoveRightTransform(-5)
    end
    
  else
    if Input.GetKeyDown(KeyCode.B) then
      self:StartConstruction('MockupEntity')
    end
  end
end

return BuilderEntity
